compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp

```diff
@@ -152,6 +152,45 @@ static int32_t getRoundedElementByteWidth(Type type) {
   return llvm::PowerOf2Ceil(byteAligned);
 }
 
+Value PackedStorageAttr::calculateStorageSizeInBytes(
+    Location loc, OpBuilder &builder, RankedTensorType type,
+    ValueRange dynamicDims) const {
+  unsigned elementBits = type.getElementTypeBitWidth();
+  assert(elementBits == 1 && "packed storage only supported for i1 for now");
+
+  // Calculate all static dims first, if any.
+  bool isPackedStorage = IREE::Encoding::hasPackedStorageAttr(type);
+  int64_t staticCount = 1;
+  if (!isPackedStorage) {
+    staticCount *= elementBits * 8;
+  }
+
+  for (unsigned i = 0; i < type.getRank(); ++i) {
+    if (!type.isDynamicDim(i))
+      staticCount *= type.getDimSize(i);
+  }
+  // Scale by dynamic dims, if present.
+  auto value =
+      builder.create<arith::ConstantIndexOp>(loc, staticCount).getResult();
+  for (auto dim : dynamicDims) {
+    value = builder.createOrFold<arith::MulIOp>(loc, value, dim);
+  }
+  // Sub-byte packing requires putting multiple elements in the same byte.
+  if (isPackedStorage) {
+    assert(8 % elementBits == 0);
+    unsigned byteElements = 8 / elementBits;
+    // TODO(antiagainst): We may want to emit runtime check to make sure this is
+    // divisible.
+    auto divisor = builder.create<arith::ConstantIndexOp>(loc, byteElements);
+    if (!isPackedStorage && dynamicDims.empty() &&
+        (staticCount * elementBits) % 8 != 0) {
+      return nullptr;
+    }
+    value = builder.createOrFold<arith::CeilDivUIOp>(loc, value, divisor);
+  }
```

Comment:
I think we can simplify this part a lot because we already assert if elementBits is 1 or not in the beginning. My mental model is keeping code simple, and extend it when needed. We definitely need to balance the mental model, but this part is much more complicated than the current need IMO.

btw, I recently learned that we prefer using CeilDivSIOp unless there is a specific reason. See https://github.com/iree-org/iree/pull/19686#discussion_r1915031028 for more details.